packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/AssetWriter.swift

```diff
@@ -0,0 +1,45 @@
+// Copyright 2013 The Flutter Authors
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import AVFoundation
+import Foundation
+
+/// A protocol that is a direct passthrough to `AVAssetWriter`. It is used to allow for mocking
+/// `AVAssetWriter` in tests.
+protocol AssetWriter: NSObjectProtocol {
+  var status: AVAssetWriter.Status { get }
+  var error: Error? { get }
+
+  func startWriting() -> Bool
+  func finishWriting(completionHandler handler: @escaping @Sendable () -> Void)
+  func startSession(atSourceTime startTime: CMTime)
+  func add(_ input: AVAssetWriterInput)
+}
+
+/// A protocol that is a direct passthrough to `AVAssetWriterInput`. It is used to allow for mocking
+/// `AVAssetWriterInput` in tests.
+protocol AssetWriterInput: NSObjectProtocol {
+  /// The underlying `AVAssetWriterInput` instance. This exists so that the input
+  /// can be extracted when adding to an AVAssetWriter.
+  var avInput: AVAssetWriterInput { get }
+
+  var expectsMediaDataInRealTime: Bool { get set }
+  var isReadyForMoreMediaData: Bool { get }
+
+  func append(_ sampleBuffer: CMSampleBuffer) -> Bool
+}
+
+/// A protocol that is a direct passthrough to `AVAssetWriterInputPixelBufferAdaptor`. It is used to
+/// allow for mocking `AVAssetWriterInputPixelBufferAdaptor` in tests.
+protocol AssetWriterInputPixelBufferAdaptor: NSObjectProtocol {
+  func append(_ pixelBuffer: CVPixelBuffer, withPresentationTime presentationTime: CMTime) -> Bool
+}
+
+extension AVAssetWriter: AssetWriter {}
+
+extension AVAssetWriterInput: AssetWriterInput {
+  var avInput: AVAssetWriterInput { self }
```

Comment:
can you give an short example why we need this? why can't we use `self` directly?

Reply:
There are places where we need to access the underlying `AVAssetWriterInput` to pass it to a method e.g. `AVCaptureSession.addInput`. We technically could cast to `AVAssetWriterInput` because, aside from tests, it's correct. However, the current implementation is safer, with the only possible failure being accessing `avInput` of a mocked `AssetWriterInput` in tests